src/tree/updater_quantile_hist.cc

```diff
@@ -549,290 +1055,113 @@ void QuantileHistMaker::Builder::InitData(const GHistIndexMatrix& gmat,
-                                            const ColumnMatrix& column_matrix,
-                                            const HistCollection& hist,
-                                            const DMatrix& fmat,
-                                            RegTree* p_tree) {
-  builder_monitor_.Start("ApplySplit");
-  // TODO(hcho3): support feature sampling by levels
-
-  /* 1. Create child nodes */
-  NodeEntry& e = snode_[nid];
-  bst_float left_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.left_sum) * param_.learning_rate;
-  bst_float right_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.right_sum) * param_.learning_rate;
-  p_tree->ExpandNode(nid, e.best.SplitIndex(), e.best.split_value,
-                     e.best.DefaultLeft(), e.weight, left_leaf_weight,
-                     right_leaf_weight, e.best.loss_chg, e.stats.sum_hess);
-
-  /* 2. Categorize member rows */
-  const auto nthread = static_cast<bst_omp_uint>(this->nthread_);
-  row_split_tloc_.resize(nthread);
-  for (bst_omp_uint i = 0; i < nthread; ++i) {
-    row_split_tloc_[i].left.clear();
-    row_split_tloc_[i].right.clear();
-  }
-  const bool default_left = (*p_tree)[nid].DefaultLeft();
-  const bst_uint fid = (*p_tree)[nid].SplitIndex();
-  const bst_float split_pt = (*p_tree)[nid].SplitCond();
-  const uint32_t lower_bound = gmat.cut.row_ptr[fid];
-  const uint32_t upper_bound = gmat.cut.row_ptr[fid + 1];
-  int32_t split_cond = -1;
-  // convert floating-point split_pt into corresponding bin_id
-  // split_cond = -1 indicates that split_pt is less than all known cut points
-  CHECK_LT(upper_bound,
-           static_cast<uint32_t>(std::numeric_limits<int32_t>::max()));
-  for (uint32_t i = lower_bound; i < upper_bound; ++i) {
-    if (split_pt == gmat.cut.cut[i]) {
-      split_cond = static_cast<int32_t>(i);
+  // prepare tasks
+  std::vector<std::pair<int32_t, size_t>> tasks;
+  for (size_t i = 0; i < nodes.size(); ++i) {
+    auto p_feature_set = column_sampler_.GetFeatureSet(nodes[i].depth);
+
+    const auto& feature_set = p_feature_set->HostVector();
+    const auto nfeature = static_cast<bst_uint>(feature_set.size());
+    for (size_t j = 0; j < nfeature; ++j) {
+      tasks.emplace_back(i, feature_set[j]);
     }
   }
 
-  const auto& rowset = row_set_collection_[nid];
-
-  Column column = column_matrix.GetColumn(fid);
-  if (column.GetType() == xgboost::common::kDenseColumn) {
-    ApplySplitDenseData(rowset, gmat, &row_split_tloc_, column, split_cond,
-                        default_left);
-  } else {
-    ApplySplitSparseData(rowset, gmat, &row_split_tloc_, column, lower_bound,
-                         upper_bound, split_cond, default_left);
-  }
+  // partial results
+  std::vector<std::pair<SplitEntry, SplitEntry>> splits(tasks.size());
+  // parallel enumeration
+  #pragma omp parallel for schedule(guided)
+  for (int32_t i = 0; i < tasks.size(); ++i) {
```

Comment:
I really like how we are evaluating both node `nid` and its sibling in the same thread. Nicely done!